decision-records/010-remove-init-method.md

```diff
@@ -0,0 +1,98 @@
+
+The `init()` method enables the following, that none of our components benefit from, so we can remove that extra step:
+
+- delaying the initialisation of the component after its instantiation
+- allowing the initialisation code to be `async` (which a `constructor` cannot be)
+
+With our intent to [throw errors at construction](./009-throw-errors-component-constructors.md), having a separate `init()` method requires careful separation between validation of the component in the `constructor()` and initialisation code in a subsequent call. By having only the constructor, the initialisation code can simply `throw` from within `constructor()` to ensure users always get functional instances of our components. 
+
+<!-- The reason for the decision and any comments or disagreements cited. -->
+
+## Risks and constraints
+
+### Reduce or delay the adoption of v5
+
+This change requires code changes from our users, that they may not have the capacity to implement.
+
+This is mitigated by the changes being quite small (removing a `.init()` call following the current instantiation of the component) and only affects a portion of our users, who use our component classes directly rather than through `initAll`.
+
+<!-- Any accepted risks associated with the decision. -->
+
+## Alternatives considered
+
+### Deprecating the init method in v5 for future removal in v6
+
+Because there wasn't really an alternative for users to switch to if we'd deprecated in v4 (if people stopped calling `init()`, there'd be no initialisation), we couldn't make the deprecation in v4.
+
+Implementing an empty `init()` to document its deprecation in v5 would only delay the update users need to make in their code. This will come at the detriment of the next release: on top of focusing on the topic of that new release, people will also have to update their JavaScript (which is the focus of v5).
+
+The scope of the change users have to make, which will be documented, as well as our continued support for v4 during a year after we ship v5 seems enough of a mitigation to directly remove the method.
+
+### Keeping the `init` method to make use of `async`
+
+`init` could be made asynchronous, which would isolate the initialisation of each component and allow to run asynchronous calls inside it.
+
+However:
+
+- Asynchronicity is not something any of our component need. Future components, like an `Autocomplete`, may involve some asynchronicity. However this asynchronicity applies to internals of the components, not their initialisation. For example, even if an Autocomplete component was relying on fetching a list of suggestion from the server, we'd want to update the UI to a text field immediately, synchronously, and delay only the display of these suggestions. Other systems (like events) could be set up to notify service code of the availability of the suggestion, rather than have it `await` the `init` call.
```

Comment:
Yeah I understand why. It was important for us, but not important for the decision

```suggestion
- Future components, like an `Autocomplete`, may add an `async init()` to  fetch a list of suggestion from the server.
```

Reply:
Sorry, I'm still pretty keen to have a bit more detail on that part, as there was a bit of confusion between asynchronous behaviour and having an asynchrounous `init()`. I've split up the example for better legibility, so you can get the gist at the start and only dive in the example if you want more detail.

```suggestion
- Asynchronicity is not something any of our components need. Future components may add asynchronous methods, but this would not make their initialisation asynchronous.

  For example, if an `Autocomplete` component fetched a list of suggestion from the server, we'd want to update the UI to a text field immediately, synchronously, and delay only the display of these suggestions. Other mechanisms (like events) could be set up to notify service code of the availability of the suggestions, rather than have it `await` an `init` call.
```

Reply:
It's still confusing to me. I'm not sure how UI delays, text field updates or events affect the decision?

That's something we can talk about in future

Reply:
Ok, I'll drop the example. We can always point back to the PR review to jog our memory 😊
```suggestion
- Asynchronicity is not something any of our components need. Future components may add asynchronous methods, but this would not require a public `async init()` method.
```

Reply:
Thanks Romaric

Can we turn that flipped _**would not require**_ wording back into a future possibility?

>The `init()` method could be made asynchronous, which would isolate the initialisation of each component and allow it to wait for asynchronous calls without blocking the main thread.
>
>However, since:
>
>- Asynchronous code requires a deeper understanding of concurrency
>- Deferred errors from Promises must be awaited before they throw
>- None of our components make asynchronous calls
>
>We can currently consider the `init()` method unnecessary.
>
>That said, we may add `async init()` methods in future. For example, an autocomplete component could immediately update a [Text input](https://design-system.service.gov.uk/components/text-input/) before asynchronously fetching a list of suggestions.

Reply:
@colinrotherham That's a neat rewrite 😍  I've updated the section with it. Just tweaked the last bit to not be too specific about which methods we may make asynchronous, just leaving the door open to our use of async methods in the future.

Reply:
Yeah course, I thought you'd want to keep your Text input example!

Thanks